superset-frontend/src/explore/components/controls/AnnotationLayerControl/AnnotationLayer.jsx

```diff
@@ -276,96 +306,224 @@ class AnnotationLayer extends React.PureComponent {
-                slice: {
-                  ...x,
-                  data: {
-                    ...x.form_data,
-                    groupby: x.form_data.groupby?.map(column =>
-                      getColumnLabel(column),
-                    ),
-                  },
-                },
-              })),
-          });
-        });
-      } else {
-        this.setState({
-          isLoadingOptions: false,
-          valueOptions: [],
-        });
-      }
+  fetchNativeAnnotations = async (search, page, pageSize) => {
+    const queryParams = rison.encode({
+      filters: [
+        {
+          col: 'name',
+          opr: 'ct',
+          value: search,
+        },
+      ],
+      columns: ['id', 'name'],
+      page,
+      page_size: pageSize,
+    });
+
+    const { json } = await SupersetClient.get({
+      endpoint: `/api/v1/annotation_layer/?q=${queryParams}`,
+    });
+
+    const { result, count } = json;
+
+    const layersArray = result.map(layer => ({
+      value: layer.id,
+      label: layer.name,
+    }));
+
+    return {
+      data: layersArray,
+      totalCount: count,
+    };
+  };
+
+  fetchCharts = async (search, page, pageSize) => {
+    const { annotationType } = this.state;
+
+    const queryParams = rison.encode({
+      filters: [
+        { col: 'slice_name', opr: 'chart_all_text', value: search },
+        {
+          col: 'id',
+          opr: 'chart_owned_created_favored_by_me',
+          value: true,
+        },
+      ],
+      columns: ['id', 'slice_name', 'viz_type'],
+      order_column: 'slice_name',
+      order_direction: 'asc',
+      page,
+      page_size: pageSize,
+    });
+    const { json } = await SupersetClient.get({
+      endpoint: `/api/v1/chart/?q=${queryParams}`,
+    });
+
+    const { result, count } = json;
+    const registry = getChartMetadataRegistry();
+
+    const chartsArray = result
+      .filter(chart => {
+        const metadata = registry.get(chart.viz_type);
+        return metadata && metadata.canBeAnnotationType(annotationType);
+      })
+      .map(chart => ({
+        value: chart.id,
+        label: chart.slice_name,
+        viz_type: chart.viz_type,
+      }));
+
+    return {
+      data: chartsArray,
+      totalCount: count,
+    };
+  };
+
+  fetchOptions = (search, page, pageSize) => {
+    const { sourceType } = this.state;
+
+    if (sourceType === ANNOTATION_SOURCE_TYPES.NATIVE) {
+      return this.fetchNativeAnnotations(search, page, pageSize);
     }
+
+    if (requiresQuery(sourceType)) {
+      return this.fetchCharts(search, page, pageSize);
+    }
+
+    return {
+      data: [],
+      totalCount: 0,
+    };
+  };
+
+  fetchSliceData = id => {
+    const queryParams = rison.encode({
+      filters: [
+        {
+          col: 'id',
+          opr: 'eq',
+          value: id,
+        },
+      ],
+      columns: ['form_data'],
+    });
+    SupersetClient.get({
+      endpoint: `/api/v1/chart/?q=${queryParams}`,
```

Comment:
Did you use the endpoint `/api/v1/chart/ `instead of `/api/v1/chart/<id>` because you can't only query for the `form_data`?

Reply:
Yes, the properties contained in `form_data` were nested and spread in the response to `/api/v1/chart/<id>` in such a way that `?q=queryParams` seemed like a more efficient method. In fact, I'm not sure how to properly query for this information according to superset guidelines, I followed the patterns already established in this component. The only doc I could find on the subject [is here](https://github.com/apache/superset/blob/master/CONTRIBUTING.md#query) after you mentioned the `columns` keyword.

Reply:
@rtexelm You can use [Superset's Swagger](https://superset.apache.org/docs/api/) to check the API. It looks like `/api/v1/chart/<id>` supports `columns` too so it would be good to call that endpoint 😉

Reply:
@michael-s-molina wow, this is exactly what I was looking for! Who would have guessed the information is precisely where it should be? haha 🫠  thank you sincerely, I was just scouring the codebase this whole time

Reply:
spoke too soon, looks like `form_data` is nested in the `query_context`. not sure how to isolate that using the `/${id}` endpoint if it's possible

Reply:
@michael-s-molina I was able to fix this, thought there was an issue with the result object but it was just my mistake. Curious if it this looks good, when you have a moment.